File: src/export/layout.rs
//! レイアウト設定モジュール
//!
//! mm基準のレイアウト定義（Source of Truth）
//! React版 layoutConfig.ts と同一の設計思想

// ============================================
// mm基準レイアウト（Source of Truth）
// ============================================

/// A4サイズ（mm）
pub const A4_WIDTH_MM: f32 = 210.0;
pub const A4_HEIGHT_MM: f32 = 297.0;

/// 余白設定（mm）
pub const MARGIN_MM: f32 = 10.0;
pub const PHOTO_GAP_MM: f32 = 10.0;

/// プレビュー比率（これが正）
pub const IMAGE_RATIO: f32 = 0.65;
pub const INFO_RATIO: f32 = 0.35;

/// 利用可能幅から写真/情報幅を計算（mm）
pub const USABLE_WIDTH_MM: f32 = A4_WIDTH_MM - MARGIN_MM * 2.0;  // 190mm
pub const PHOTO_WIDTH_MM: f32 = USABLE_WIDTH_MM * IMAGE_RATIO;   // 123.5mm
pub const INFO_WIDTH_MM: f32 = USABLE_WIDTH_MM * INFO_RATIO;     // 66.5mm

/// 写真高さ: 3枚配置から計算（mm）
pub const PHOTO_HEIGHT_MM_3UP: f32 = (A4_HEIGHT_MM - MARGIN_MM * 2.0 - PHOTO_GAP_MM * 2.0) / 3.0;  // 85.67mm
pub const PHOTO_HEIGHT_MM_2UP: f32 = (A4_HEIGHT_MM - MARGIN_MM * 2.0 - PHOTO_GAP_MM) / 2.0;        // 128.5mm

// ============================================
// 変換係数
// ============================================

/// mm → pt変換 (1mm = 72/25.4 pt ≈ 2.835pt)
pub const MM_TO_PT: f32 = 72.0 / 25.4;

/// pt → px変換 (96dpi基準)
pub const PT_TO_PX: f32 = 96.0 / 72.0;
pub const PX_TO_PT: f32 = 72.0 / 96.0;

// ============================================
// 導出されるpt値
// ============================================

/// ページサイズ（pt）
pub const PAGE_WIDTH_PT: f32 = A4_WIDTH_MM * MM_TO_PT;   // 595.35pt
pub const PAGE_HEIGHT_PT: f32 = A4_HEIGHT_MM * MM_TO_PT; // 842.0pt

/// マージン（pt）
pub const MARGIN_PT: f32 = MARGIN_MM * MM_TO_PT;         // 28.35pt
pub const GAP_PT: f32 = PHOTO_GAP_MM * MM_TO_PT;         // 28.35pt

/// 写真サイズ（pt）
pub const PHOTO_WIDTH_PT: f32 = PHOTO_WIDTH_MM * MM_TO_PT;       // 350.1pt
pub const PHOTO_HEIGHT_PT_3UP: f32 = PHOTO_HEIGHT_MM_3UP * MM_TO_PT; // 242.9pt
pub const PHOTO_HEIGHT_PT_2UP: f32 = PHOTO_HEIGHT_MM_2UP * MM_TO_PT; // 364.3pt

/// 情報パネル幅（pt）
pub const INFO_WIDTH_PT: f32 = INFO_WIDTH_MM * MM_TO_PT;         // 188.5pt

// ============================================
// フィールド定義
// ============================================

/// 情報パネルに表示するフィールド
#[derive(Debug, Clone, Copy)]
pub struct FieldDefinition {
    pub key: &'static str,
    pub label: &'static str,
    pub row_span: u8,
}

/// レイアウトフィールド（React版 LAYOUT_FIELDS と同等）
pub const LAYOUT_FIELDS: &[FieldDefinition] = &[
    FieldDefinition { key: "workType", label: "工種", row_span: 1 },
    FieldDefinition { key: "variety", label: "種別", row_span: 1 },
    FieldDefinition { key: "detail", label: "細別", row_span: 1 },
    FieldDefinition { key: "station", label: "測点", row_span: 1 },
    FieldDefinition { key: "remarks", label: "備考", row_span: 2 },
    FieldDefinition { key: "measurements", label: "測定値", row_span: 3 },
];

// ============================================
// レイアウト設定構造体
// ============================================

/// PDFレイアウト設定
#[derive(Debug, Clone)]
pub struct PdfLayout {
    /// ページ幅（mm）
    pub page_width_mm: f32,
    /// ページ高さ（mm）
    pub page_height_mm: f32,
    /// マージン（mm）
    pub margin_mm: f32,
    /// 写真間ギャップ（mm）
    pub gap_mm: f32,
    /// 写真幅（mm）
    pub photo_width_mm: f32,
    /// 写真高さ（mm）
    pub photo_height_mm: f32,
    /// 情報パネル幅（mm）
    pub info_width_mm: f32,
    /// 1ページあたりの写真数
    pub photos_per_page: u8,
}

impl PdfLayout {
    /// 3枚/ページ用レイアウト
    pub fn three_up() -> Self {
        Self {
            page_width_mm: A4_WIDTH_MM,
            page_height_mm: A4_HEIGHT_MM,
            margin_mm: MARGIN_MM,
            gap_mm: PHOTO_GAP_MM,
            photo_width_mm: PHOTO_WIDTH_MM,
            photo_height_mm: PHOTO_HEIGHT_MM_3UP,
            info_width_mm: INFO_WIDTH_MM,
            photos_per_page: 3,
        }
    }

    /// 2枚/ページ用レイアウト
    pub fn two_up() -> Self {
        Self {
            page_width_mm: A4_WIDTH_MM,
            page_height_mm: A4_HEIGHT_MM,
            margin_mm: MARGIN_MM,
            gap_mm: PHOTO_GAP_MM,
            photo_width_mm: PHOTO_WIDTH_MM,
            photo_height_mm: PHOTO_HEIGHT_MM_2UP,
            info_width_mm: INFO_WIDTH_MM,
            photos_per_page: 2,
        }
    }

    /// 指定枚数でレイアウト取得
    pub fn for_photos_per_page(n: u8) -> Self {
        match n {
            2 => Self::two_up(),
            _ => Self::three_up(),
        }
    }

    /// ブロック高さ（写真 + ギャップ）mm
    pub fn block_height_mm(&self) -> f32 {
        self.photo_height_mm + self.gap_mm
    }

    /// 利用可能幅（mm）
    pub fn usable_width_mm(&self) -> f32 {
        self.page_width_mm - self.margin_mm * 2.0
    }

    /// 利用可能高さ（mm）
    pub fn usable_height_mm(&self) -> f32 {
        self.page_height_mm - self.margin_mm * 2.0
    }

    /// コンテンツ開始Y座標（mm、上から）
    pub fn content_start_y_mm(&self) -> f32 {
        self.page_height_mm - self.margin_mm
    }
}

// ============================================
// ヘルパー関数
// ============================================

/// mm → pt 変換
#[inline]
pub fn mm_to_pt(mm: f32) -> f32 {
    mm * MM_TO_PT
}

/// pt → mm 変換
#[inline]
pub fn pt_to_mm(pt: f32) -> f32 {
    pt / MM_TO_PT
}

/// px → pt 変換
#[inline]
pub fn px_to_pt(px: f32) -> f32 {
    px * PX_TO_PT
}

/// pt → px 変換
#[inline]
pub fn pt_to_px(pt: f32) -> f32 {
    pt * PT_TO_PX
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_dimensions() {
        // 基本寸法の確認
        assert!((USABLE_WIDTH_MM - 190.0).abs() < 0.01);
        assert!((PHOTO_WIDTH_MM - 123.5).abs() < 0.01);
        assert!((INFO_WIDTH_MM - 66.5).abs() < 0.01);
        assert!((PHOTO_HEIGHT_MM_3UP - 85.67).abs() < 0.1);
    }

    #[test]
    fn test_ratios() {
        // 比率の確認
        let total = IMAGE_RATIO + INFO_RATIO;
        assert!((total - 1.0).abs() < 0.001);
    }

    #[test]
    fn test_conversion() {
        // 変換係数の確認
        assert!((MM_TO_PT - 2.835).abs() < 0.01);
        assert!((mm_to_pt(10.0) - 28.35).abs() < 0.1);
    }

    #[test]
    fn test_layout_config() {
        let layout = PdfLayout::three_up();
        assert_eq!(layout.photos_per_page, 3);
        assert!((layout.photo_width_mm - 123.5).abs() < 0.01);

        let layout2 = PdfLayout::two_up();
        assert_eq!(layout2.photos_per_page, 2);
        assert!(layout2.photo_height_mm > layout.photo_height_mm);
    }
}

File: photo-ai-go/internal/export/layout.go
// Package export provides layout constants and export orchestration for PDF/Excel generation.
//
// Constants are ported from common/src/layout.rs (GAS-compliant layout).
package export

// A4 page dimensions in points (GAS-compliant).
const (
	A4WidthPt  float64 = 595.35
	A4HeightPt float64 = 842.0

	A4WidthMM  float64 = 210.0
	A4HeightMM float64 = 297.0
)

// A4 landscape dimensions in points.
const (
	A4LandscapeWidthPt  float64 = 842.0
	A4LandscapeHeightPt float64 = 595.35
)

// Layout spacing in points (GAS-compliant).
const (
	MarginPt float64 = 28.35 // 10mm
	GapPt    float64 = 5.0
)

// Photo/info area ratios (GAS-compliant).
const (
	ImageRatio float64 = 0.65
	InfoRatio  float64 = 0.35
)

// Header height in points (GAS-compliant: no header).
const HeaderHeightPt float64 = 0.0

// Pair PDF margins/dimensions in points.
const (
	PairMarginYPt      float64 = 28.35
	PairMarginXPt      float64 = 19.2
	PairCaptionAreaPt  float64 = 43.2
	PairRuleWidthPt    float64 = 264.0
)

// mm-based constants (Excel use).
const (
	MarginMM   float64 = 10.0
	PhotoGapMM float64 = 10.0

	UsableWidthMM float64 = A4WidthMM - MarginMM*2   // 190mm
	PhotoWidthMM  float64 = UsableWidthMM * ImageRatio // 123.5mm
	InfoWidthMM   float64 = UsableWidthMM * InfoRatio  // 66.5mm

	PhotoAspectRatio float64 = 4.0 / 3.0
	PhotoHeightMM    float64 = PhotoWidthMM / PhotoAspectRatio // 92.625mm

	// Legacy page-division heights (kept for compatibility).
	PhotoHeightMM3Up float64 = (A4HeightMM - MarginMM*2 - PhotoGapMM*2) / 3 // 85.67mm
	PhotoHeightMM2Up float64 = (A4HeightMM - MarginMM*2 - PhotoGapMM) / 2   // 128.5mm
)

// Unit conversion factors.
const (
	MMToPt float64 = 72.0 / 25.4 // ≈ 2.835
	PtToPx float64 = 96.0 / 72.0
	PxToPt float64 = 72.0 / 96.0

	PtPerExcelCol     float64 = 5.3
	PxPerExcelCol     float64 = 7.1
	ExcelColOffsetPx  float64 = 5.0
)

// Derived pt values.
const (
	PageWidthPt  float64 = A4WidthPt
	PageHeightPt float64 = A4HeightPt

	PhotoWidthPt    float64 = PhotoWidthMM * MMToPt
	PhotoHeightPt3Up float64 = PhotoHeightMM3Up * MMToPt
	PhotoHeightPt2Up float64 = PhotoHeightMM2Up * MMToPt

	InfoWidthPt float64 = InfoWidthMM * MMToPt

	BlockHeight3UpPt float64 = PhotoHeightPt3Up + GapPt*MMToPt
)

// Excel style constants.
const (
	LabelFontSize    float64 = 9.0
	LabelFontColor   uint32  = 0x555555
	LabelBgColor     uint32  = 0xF5F5F5
	LabelBorderColor uint32  = 0xAAAAAA
	ValueFontSize    float64 = 11.0
	CellBorderColor  uint32  = 0xCCCCCC
)

// Excel layout constants.
const (
	ExcelScale float64 = 1.1

	PhotoRows3Up      uint8   = 10
	PhotoRows2Up      uint8   = 16
	GapRows           uint8   = 1
	RowsPerBlock3Up   uint8   = PhotoRows3Up + GapRows  // 11
	RowsPerBlock2Up   uint8   = PhotoRows2Up + GapRows  // 17

	RowHeightPt    float64 = 26.0
	PhotoColWidth  float64 = 56.1
	LabelColWidth  float64 = 11.0
	ValueColWidth  float64 = 28.6
	InfoColWidth   float64 = 39.6
)

// FieldKey identifies an information panel field.
type FieldKey int

const (
	FieldDate FieldKey = iota
	FieldPhotoCategory
	FieldWorkType
	FieldVariety
	FieldSubphase
	FieldStation
	FieldRemarks
	FieldMeasurements
)

// FieldDefinition describes one field in the information panel.
type FieldDefinition struct {
	Key     FieldKey
	Label   string
	RowSpan uint8
}

// LayoutFields lists the 8 information panel fields in GAS-compliant order.
var LayoutFields = []FieldDefinition{
	{FieldDate, "日時", 1},
	{FieldPhotoCategory, "区分", 1},
	{FieldWorkType, "工種", 1},
	{FieldVariety, "種別", 1},
	{FieldSubphase, "細別", 1},
	{FieldStation, "測点", 1},
	{FieldRemarks, "備考", 2},
	{FieldMeasurements, "測定値", 2},
}

// PdfLayout holds PDF page layout configuration.
type PdfLayout struct {
	PageWidthMM  float64
	PageHeightMM float64
	MarginMM     float64
	GapMM        float64
	PhotoWidthMM float64
	PhotoHeightMM float64
	InfoWidthMM  float64
	PhotosPerPage int
}

// ThreeUp returns a 3-photos/page PDF layout.
func ThreeUp() PdfLayout {
	return PdfLayout{
		PageWidthMM:   A4WidthMM,
		PageHeightMM:  A4HeightMM,
		MarginMM:      MarginMM,
		GapMM:         PhotoGapMM,
		PhotoWidthMM:  PhotoWidthMM,
		PhotoHeightMM: PhotoHeightMM3Up,
		InfoWidthMM:   InfoWidthMM,
		PhotosPerPage: 3,
	}
}

// TwoUp returns a 2-photos/page PDF layout.
func TwoUp() PdfLayout {
	return PdfLayout{
		PageWidthMM:   A4WidthMM,
		PageHeightMM:  A4HeightMM,
		MarginMM:      MarginMM,
		GapMM:         PhotoGapMM,
		PhotoWidthMM:  PhotoWidthMM,
		PhotoHeightMM: PhotoHeightMM2Up,
		InfoWidthMM:   InfoWidthMM,
		PhotosPerPage: 2,
	}
}

// ForPhotosPerPage returns the layout for n photos per page.
func ForPhotosPerPage(n int) PdfLayout {
	if n == 2 {
		return TwoUp()
	}
	return ThreeUp()
}

// BlockHeightMM returns the height of one photo block (photo + gap) in mm.
func (l PdfLayout) BlockHeightMM() float64 {
	return l.PhotoHeightMM + l.GapMM
}

// UsableWidthMM returns the printable width in mm.
func (l PdfLayout) UsableWidthMM() float64 {
	return l.PageWidthMM - l.MarginMM*2
}

// UsableHeightMM returns the printable height in mm.
func (l PdfLayout) UsableHeightMM() float64 {
	return l.PageHeightMM - l.MarginMM*2
}

// ExcelLayout holds Excel workbook layout configuration.
type ExcelLayout struct {
	RowsPerBlock  uint8
	PhotoRows     uint8
	RowHeightPt   float64
	ColAWidth     float64
	ColBWidth     float64
	ColCWidth     float64
	PhotoWidthPt  float64
	PhotoHeightPt float64
	InfoWidthPt   float64
}

// ExcelThreeUp returns a 3-photos/page Excel layout.
func ExcelThreeUp() ExcelLayout {
	return ExcelLayout{
		RowsPerBlock:  RowsPerBlock3Up,
		PhotoRows:     PhotoRows3Up,
		RowHeightPt:   RowHeightPt,
		ColAWidth:     PhotoColWidth,
		ColBWidth:     LabelColWidth,
		ColCWidth:     ValueColWidth,
		PhotoWidthPt:  PhotoWidthPt,
		PhotoHeightPt: PhotoHeightPt3Up,
		InfoWidthPt:   InfoWidthPt,
	}
}

// ExcelTwoUp returns a 2-photos/page Excel layout.
func ExcelTwoUp() ExcelLayout {
	return ExcelLayout{
		RowsPerBlock:  RowsPerBlock2Up,
		PhotoRows:     PhotoRows2Up,
		RowHeightPt:   RowHeightPt,
		ColAWidth:     PhotoColWidth,
		ColBWidth:     LabelColWidth,
		ColCWidth:     ValueColWidth,
		PhotoWidthPt:  PhotoWidthPt,
		PhotoHeightPt: PhotoHeightPt2Up,
		InfoWidthPt:   InfoWidthPt,
	}
}

// ExcelForPhotosPerPage returns the Excel layout for n photos per page.
func ExcelForPhotosPerPage(n int) ExcelLayout {
	if n == 2 {
		return ExcelTwoUp()
	}
	return ExcelThreeUp()
}

// PhotoAreaPx returns the photo area dimensions in pixels.
func (l ExcelLayout) PhotoAreaPx() (width, height uint32) {
	rowHeightPx := uint32(l.RowHeightPt * PtToPx)
	w := uint32(l.ColAWidth*PxPerExcelCol + ExcelColOffsetPx)
	h := rowHeightPx * uint32(l.PhotoRows)
	return w, h
}

// MmToPt converts millimetres to points.
func MmToPt(mm float64) float64 { return mm * MMToPt }

// PtToMm converts points to millimetres.
func PtToMm(pt float64) float64 { return pt / MMToPt }

// FitImageCentered computes scaled dimensions and offsets to fit an image
// centred inside a target rectangle while preserving aspect ratio.
// Returns (scaledW, scaledH, offsetX, offsetY).
func FitImageCentered(imageW, imageH, targetW, targetH float64) (float64, float64, float64, float64) {
	scale := min64(targetW/imageW, targetH/imageH)
	scaledW := imageW * scale
	scaledH := imageH * scale
	offsetX := max64((targetW-scaledW)/2, 0)
	offsetY := max64((targetH-scaledH)/2, 0)
	return scaledW, scaledH, offsetX, offsetY
}

func min64(a, b float64) float64 {
	if a < b {
		return a
	}
	return b
}

func max64(a, b float64) float64 {
	if a > b {
		return a
	}
	return b
}
